### ISSUE TITLE
Endless CPU and memory occur in "node" when a symlink to / exists in the workspace

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode#121854]
- VS Code Version: 1.55.2 and the today's Insiders.
- OS Version: Ubuntu 20 (via WSL2).

Steps to Reproduce:

1. Go to some folder via the command line.
2. ln -s / ./test1
3. Open the folder in VS Code.

Bug: Node starts using 200% CPU and memory keeps increasing (slowly). Even adding test1 to files.exclude doesn't work. Using the Toggle Developer Tools->Performance logging shows no CPU usage.

Does this issue occur when all extensions are disabled?: Yes

**Comment by deepak1556** [comment]
Can you configure to exclude the folder under `files.watcherExclude` and see if the issue happens.

**Comment by sean-mcmanus** [comment]
That does not fix it.

**Comment by deepak1556** [comment]
Can you provide the output of `cat /proc/<PID>/cmdline` where PID is of the process with high cpu usage.

**Comment by sean-mcmanus** [comment]
/home/username/.vscode-server-insiders/bin/85f8ebf1687661c45fe2633bdd7f640164a20cc2/node/home/username/.vscode-server-insiders/bin/85f8ebf1687661c45fe2633bdd7f640164a20cc2/out/bootstrap-fork--type=watch

**Comment by deepak1556** [comment]
Thanks, what is the version of remote wsl extension being used ? Can you check if the issue goes away with `"remote.WSL.fileWatcher.polling": false`

**Comment by sean-mcmanus** [comment]
0.56.1. remote.WSL.fileWatcher.polling is already set to false by default and explicitly setting it to false has no effect.

**Comment by sean-mcmanus** [comment]
So it sounds like this is a Remote/WSL2-only issue? I could try the repro on a local Linux machine if you want.

**Comment by deepak1556** [comment]
> I could try the repro on a local Linux machine if you want.

Yes that would be helpful if you can do it.

**Comment by sean-mcmanus** [comment]
Yeah, I don't repro it on Linux arm64 machine when files.watcherExclude is set, but it repros when I set the path to false (which sounds like it's "by design", although it seems odd  why files.watcherExclude doesn't inherit files.exclude like search.exclude does). 

So it looks like the bug is that files.watcherExclude isn't working in the remote WSL scenario (unless there's some other factor I'm not aware of).

**Comment by bpasero** [comment]
@sean-mcmanus still repros in insiders?

**Comment by sean-mcmanus** [comment]
@bpasero This bug still repros with today's Insiders.

**Comment by bpasero** [comment]
@sean-mcmanus can you set log level to Trace:

![image](https://user-images.githubusercontent.com/900690/135964101-bc17bc68-92a4-46ee-914d-e2684342387a.png)

And open the server output channel, open a file, make some changes, save and see what output you get from the file watcher. I am curious as to what file watcher is being used in your case.

Should be prefixed with something like `[trace] [File Watcher ...`

**Comment by sean-mcmanus** [comment]
@bpasero It says [remoteagent] [trace] [File Watcher (node.js)].

**Comment by bpasero** [comment]
@sean-mcmanus the `node.js` watcher is a specific one only being used for watching files that are not inside the workspace. There should be other messages including `nsfw` which is the watcher we use for deep recursive folder watching.

Anyway, I just tried this out myself both on WSL1 and WSL2 and already found 2 issues:
* the process explorer (from help) menu does not seem to reflect the high CPU usages (https://github.com/microsoft/vscode/issues/134475)
* I see a large spike in `ripgrep` running for a long time with many processes (https://github.com/microsoft/vscode/issues/134476)

As for the file watcher process: I see it running and I see it consuming around 20% of CPU (that is what `htop` reports for me). It is somewhat expected that the watcher consumes CPU at least on startup on Linux but it will eventually settle down: Linux does not have efficient recursive file watching, so a watcher will have to recursively resolve all child folders of a certain root and install itself as watcher. I suspect this is causing the higher CPU load on startup from the watcher.

Anyhow, once we introduce a different file watcher, I suspect this might change because the new file watcher will no longer automatically follow symbolic links when encountered. Instead, you have to explicitly configure `files.watcherInclude` to add such symbolic links.
